Set extra proxy headers in all `tsh` HTTP requests

vitorenesduarte opened the pull request:
Fixes #16541.

Before this commit, the `tsh` HTTP requests that had the extra headers were those that did not use `roundtrip`.
This commit leverages `http.RoundTripper.RoundTrip` to ensure that all requests have the the extra headers.

Commit:
Set extra proxy headers in all `tsh` HTTP requests

Before this commit, the `tsh` HTTP requests that had the extra headers
were those that did not use `roundtrip`.
This commit leverages `http.RoundTripper.RoundTrip` to ensure that all
requests have the the extra headers.

ibeckermayer commented on lib/client/weblogin.go:
```diff
@@ -360,7 +374,13 @@ func SSHAgentSSOLogin(ctx context.Context, login SSHLoginSSO, config *Redirector
 // SSHAgentLogin is used by tsh to fetch local user credentials.
 func SSHAgentLogin(ctx context.Context, login SSHLoginDirect) (*auth.SSHLoginResponse, error) {
-	clt, _, err := initClient(login.ProxyAddr, login.Insecure, login.Pool)
+	cltConfig := ClientConfig{
+		ProxyAddr:    login.ProxyAddr,
+		Insecure:     login.Insecure,
+		Pool:         login.Pool,
+		ExtraHeaders: login.ExtraHeaders,
+	}
+	clt, _, err := initClient(cltConfig)
```
```suggestion
clt, _, err := initClient(ClientConfig{
		ProxyAddr:    login.ProxyAddr,
		Insecure:     login.Insecure,
		Pool:         login.Pool,
		ExtraHeaders: login.ExtraHeaders,
	})
```

ibeckermayer reviewed (changes requested):
Some tests testing the logic of whether requests get downgraded or not would be nice.

Also try a `grep -rn NewInsecureWebClient\(\)`, there seem to be some stragglers left over.

Commit:
Update http round tripper docs

Co-authored-by: Isaiah Becker-Mayer <isaiah@goteleport.com>

vitorenesduarte commented on api/client/proxy/proxy.go:
```diff
@@ -71,32 +71,39 @@ func parse(addr string) (*url.URL, error) {
-// HTTPFallbackRoundTripper is a wrapper for http.Transport that downgrades requests
-// to plain HTTP when using a plain HTTP proxy at localhost.
-type HTTPFallbackRoundTripper struct {
+// HTTPRoundTripper is a wrapper for http.Transport that
```
I renamed it because now it has a double purpose (downgrading and adding http headers).

Commit:
Inline config definition

Commit:
Add `NewInsecureWebClient` back

Commit:
Merge branch 'master' into vitor/extra-proxy-headers

Commit:
Fix TestProxyAwareRoundTripper

Commit:
Merge branch 'vitor/extra-proxy-headers' of https://github.com/gravitational/teleport into vitor/extra-proxy-headers

Commit:
Merge branch 'master' into vitor/extra-proxy-headers

Commit:
Don't change config state

vitorenesduarte commented on lib/client/weblogin.go:
```diff
@@ -249,39 +251,51 @@ type SSHLoginPasswordless struct {
 	}
-	proxyAddr = "https://" + net.JoinHostPort(host, port)
-	u, err := url.Parse(proxyAddr)
+	config.ProxyAddr = "https://" + net.JoinHostPort(host, port)
```
Done in c1514a4a27e3b1d02e98f4c1c2e7cbbbcf7fabf9.

Commit:
Remove unnecessary path var in TestPlainHttpFallback

Commit:
Remove ClientConfig struct

Commit:
Merge branch 'master' into vitor/extra-proxy-headers

Commit:
Rename tests

Commit:
Simplify TestPlainHttpFallback

Commit:
Add http roundtripper tests

Commit:
Merge branch 'master' into vitor/extra-proxy-headers

vitorenesduarte commented:
Thanks for the review @ibeckermayer. Could you please take another look?

> Some tests testing the logic of whether requests get downgraded or not would be nice.

I improved existing tests and added new ones in 9641dd13aaaf84e89e7b2093e1c1c127c7531fc0.

> Also try a `grep -rn NewInsecureWebClient\(\)`, there seem to be some stragglers left over.

Thanks. I've added `NewInsecureWebClient` back.

Commit:
Improve comment

ibeckermayer commented on lib/client/weblogin_test.go:
```diff
@@ -31,74 +33,257 @@ import (
-				defer httpSvr.Close()
+// TestHttpRoundTripperDowngrade tests that the round tripper downgrades https requests to http
+// when HTTP_PROXY is set to "http://localhost:*" (i.e. there's an http proxy running on localhost).
+func TestHttpRoundTripperDowngrade(t *testing.T) {
```
I get that this is testing the `HttpRoundTripper` under the hood, however its bad practice to write tests for one package in another. For example, consider if we swap out the underlying transport mechanism in `SSHAgentLogin`. Now this test is bunk, and might just get deleted. At the very least, we would need to rewrite it to test `HTTPRoundTripper` in its own package at that point. 

Another reason this is bad practice is that tests are a form of documentation. If somebody wants to use `HTTPRoundTripper` for something else, it's best that the tests testing its core behavior are close by so they can see how it works.

I would say move these comprehensive tests of `HTTPRoundTripper`'s behavior close to `TestProxyAwareRoundTripper`. And if you think its worthwhile to also have unit tests ensuring this behavior for `client.SSHAgentLogin` and `client.HostCredentials`, then add those to the `client` package.

vitorenesduarte replied:
Thanks for the suggestion. Moved the tests in 451646b116a2c646071d235ec78cd9c3717d11d6.

Commit:
Remove optimization in RoundTrip

vitorenesduarte commented on api/client/proxy/proxy.go:
```diff
@@ -71,32 +71,42 @@ func parse(addr string) (*url.URL, error) {
+	// extraHeaders is a map of extra HTTP headers to be included in requests.
+	extraHeaders map[string]string
+	// downgrade indicates that requests to plain HTTP should be downgraded.
+	downgrade bool
```
Good point, thanks! Reverted it in 0c3a5dd8f1fb173b22a991176bd19a377168e066.

Commit:
Improve TestHttpRoundTripperDowngrade

vitorenesduarte commented on lib/client/weblogin_test.go:
```diff
@@ -31,74 +33,257 @@ import (
+			require.NoError(t, err)
+		})
+	}
+}
```
Done in aa639b1b4d10f2af3f5f963249105e815497c7bc.

Commit:
Move TestHttpRoundTripper* tests to api/client/proxy

Commit:
Merge branch 'master' into vitor/extra-proxy-headers

Commit:
Merge branch 'master' into vitor/extra-proxy-headers

Commit:
Merge branch 'master' into vitor/extra-proxy-headers

Commit:
Merge branch 'master' into vitor/extra-proxy-headers

Commit:
Don't use roundtrip in roundtripper tests

Commit:
Fix lint-api

Commit:
Consistent comments

Commit:
Merge branch 'master' into vitor/extra-proxy-headers

Commit:
Merge branch 'master' into vitor/extra-proxy-headers

Commit:
Merge branch 'master' into vitor/extra-proxy-headers

vitorenesduarte commented:
Manually backported to v11: #20071
Manually backporting to < v11 resulted in conflicts due to #15874 not being on those branches.
